### ISSUE TITLE
Please integrate with GitLab.com

### ISSUE THREAD

**Issue by thozza** [docker/hub-beta-feedback#334]
Hi.

Since Gitlab already provides their hosted service for git repositories it would be great to have Docker Hub being integrated with Gitlab.com. Right now I have to mirror my repo with Dockerfile on Github or Bitbucket to be able to have automated builds.

**Comment by kencochrane** [comment]
It is already in progress :)

**Comment by thozza** [comment]
@kencochrane great! Thanks for the info. :wink:

**Comment by anandchida** [comment]
Do we have an ETA of this integration? Would love to move to hosted docker hub if gitlab support is available

**Comment by Jonjoe** [comment]
+1 on giving an eta :) we have had to split our repos between bb and github. If we can get better GL integration it would be a dream come true <3

**Comment by kencochrane** [comment]
Paging @mattsoldo so he can give an eta.

**Comment by joeldrapper** [comment]
+1

**Comment by ejaz-ahmed** [comment]
Any update on this issue?

**Comment by pgrm** [comment]
:+1:

**Comment by jangorecki** [comment]
+1

**Comment by zwarag** [comment]
@kencochrane Will this work with private hosted GitLabs to or only with the main one?
Or same question just different, do you plan on adding support for plain git like quay.io does?

**Comment by kencochrane** [comment]
@pchico83 @dustinlacewell @caervs Can you answer those questions?

**Comment by pkennedyr** [comment]
Thanks for the feedback! We are actively expanding our list of supported source providers on both Docker Hub and Docker Cloud and are targeting support for GitLab later in 2016.

Re: linking to a custom Git repository, we are happy to explore this option as well. Please upvote here to indicate interest in this feature, which will help us prioritize accordingly.

**Comment by jimador** [comment]
+1 It would be nice to have Gitlab integration, but a custom git solution might be a more generic solution.

**Comment by thozza** [comment]
I guess we don't need Docker Hub for GitLab  any more... https://about.gitlab.com/2016/05/23/gitlab-container-registry/ ;)

**Comment by c33s** [comment]
would prefer to use docker.io with gitlabs registry. 
but dockerhub should also support generic git repos

**Comment by DrChr** [comment]
Any update on this issue?  (An earlier post mentioned later in 2016, and now we're in a very early 2017)

**Comment by Czende** [comment]
+1

**Comment by andreasunterhuber** [comment]
+1

**Comment by teodorescuserban** [comment]
+1
any update on this?

**Comment by hraban** [comment]
@c33s gitlab CI offers automatic building and uploading to the registry, much like dockerhub with github integration. it's done by putting the regular build and push commands in a CI dot-file in the repository instead of magically done by the registry on commit, but the end result is the same: push to gitlab ---> automatically update the image in the registry.

see https://about.gitlab.com/2016/05/23/gitlab-container-registry/ ("Use with Gitlab CI" at the bottom)

I just tested this and it works fine.

**Comment by Jonjoe** [comment]
Thanks @hraban

**Comment by TheYorkshireDev** [comment]
Any update on this? As some of the previous commenters have mentioned this was opened in 2015 with the feature to be added in 2016. It is now 2017 and there is still no real news. It would be great to have gitlab as an option for been a linked account just like github and bitbucket is.

<img width="1323" alt="screen shot 2017-07-12 at 20 02 37" src="https://user-images.githubusercontent.com/30004860/28135150-1eed0552-673d-11e7-87ce-f862444e01ac.png">


Does anyone know if this is still in the works?

**Comment by mileo** [comment]
up

**Comment by murbano83** [comment]
@kencochrane @pkennedyr @pchico83 no news on this?

**Comment by metalcamp** [comment]
Any news on this? For now I figured out a solution similar to @hraban, but complete integration would be great.

**Comment by RAYs3T** [comment]
@kencochrane Anything new on this?

**Comment by teodorescuserban** [comment]
FYI my `.gitlab-ci.yml` (using a private gitlab runner) that will start building everytime a branch is changed.

```
variables:
  IMAGE_TAG: $CI_REGISTRY_IMAGE:$CI_COMMIT_REF_NAME

before_script:
  - docker login -u gitlab-ci-token -p $CI_JOB_TOKEN $CI_REGISTRY

build:
  stage: build
  script:
    - docker build -t $IMAGE_TAG .
    - docker push $IMAGE_TAG
```

**Comment by mckaygerhard** [comment]
docker cloud will no more since 21 may! ja! so in any case never was sense in "made inside other"!

**Comment by lukad** [comment]
The missing GitLab integration is the only thing keeping me from moving all my projects to GitLab.
So +1 for this!

**Comment by teodorescuserban** [comment]
In the recent events around Microsoft acquiring GitHub I do hope this issue gets some more attention.

**Comment by bbashy** [comment]
@teodorescuserban Tell me about it. Just moved everything and found out I can't change the repo to GitLab's... confusing. May just remove the project and build it in GitLab registry for myself instead of publicly.

**Comment by teodorescuserban** [comment]
@bbashy yep, thats what i did. (see above a very simple `.gitlab-ci.yml`)

**Comment by neechbear** [comment]
Any progress on this? I imagine there will be many more people who are interested in this now since the GitHub announcement.

**Comment by revof11** [comment]
I'm running with the `.gitlab-ci.yml` setup for now.  I'm also waiting for CircleCI.com to add GitLab support  for the rest of my projects.  GitLab's CI is OK but it's just not wonderful-fantastic.

https://circleci.com/ideas/?idea=CCI-I-248

**Comment by FrankGiesecke** [comment]
+1

**Comment by createchange** [comment]
+1

**Comment by Queuecumber** [comment]
@kencochrane 3 years later and no official statement, can you tell us anything?

**Comment by kencochrane** [comment]
@Queuecumber sorry I don't work for Docker anymore, so I can't help you.

**Comment by Yajo** [comment]
Gitlab ships a fully integrated docker registry, FYI.

**Comment by FrankGiesecke** [comment]
@Yajo:  That’s right, but for people who will host images on docker hub (using the automated build feature), it’s necessary to get a connector similar to GitHub.

**Comment by RAYs3T** [comment]
@FrankGiesecke You can use GitLab's build-in CI to build and push the image. See this project for example: https://gitlab.com/RAYs3T/sourcemod-plugins-builder-ci/blob/master/.gitlab-ci.yml

**Comment by FrankGiesecke** [comment]
@RAYs3T That’s also right. But using the built in CI of GitLab will cost runner time.
Automated builds on Docker Hub won’t.

Currently, the most image projects are hosted on GitHub using the automated build on Docker Hub.
Since Microsoft has taken over, many developers are looking for alternatives like GitLab. So a connector to GitLab will make sense for developers who want to migrate their projects.

**Comment by 06kellyjac** [comment]
Additionally, I don't know if this is fact but I've personally experienced automated builds ranking higher on Docker Hub searches than non-automated builds of similar or more downloads & stars.

**Comment by RAYs3T** [comment]
@FrankGiesecke You will get 2000 free minutes per month on gitlab. Even for closed projects. And the quota is calculated also per group. So you won't use your personal quota if the project is member of a group.
And 2000 minutes / month should be enough?

It seems like the gitlab integration is not really wanted - we waiting for YEARS now, and nothing is happening. So we need  some work-a-rounds for it...

@06kellyjac Ok, that one is quite important. So images  which are not costing mony (because the build-process is external) are treated worse then the pricey ones, running on the own infrastructure ❓

**Comment by 06kellyjac** [comment]
@RAYs3T It's hard to find examples but if you search `mutt` on docker hub `jessfraz/mutt` is below `cmoultrie/mutt`

| Name         | `jessfraz/mutt` | `cmoultrie/mutt` |
|--------------|-----------------|------------------|
| Automated    | No              | Yes              |
| Stars        | 0               | 0                |
| Pulls        | 187             | 138              |
| Size        | 24MB             | 123MB              |
| Last changed | 5 days ago      | 1 year ago       |

There are a couple behind the scenes factors that could maybe impact it's position in the rankings:
* More clicks on a docker image listing making it rank higher but lower pulls
* More account views or total views on other docker files from the account?
* Potential spam account???? - Jessie Frazelle's Docker Hub account is `jess` idk if it's a new account or another jessfraz but that was another thought I had

The best way to test this would probably be to create 2 new accounts & matching dockerfiles and play with staring and increasing the pull count of each docker image etc.

Note: I wouldn't be surprised if people could find cases showing the opposite. I think I have seen the opposite but it's more often not in my (limited) experience. Would be interesting to find out if that really impacts the search rankings but as I generally prefer Auto builds myself as I can look at the docker image in the Hub or click the Repo link to have a look; I don't have to pull the docker image and investigate it on my machine to see if the contents match the repo, or build the image from the dockerfile myself, as I know it was actually built from the repo's dockerfile by docker.

**Comment by Yajo** [comment]
Another option Gitlab has is to define a mirror.

**Comment by 06kellyjac** [comment]
@Yajo yep, that's mentioned in the initial issue from 2015 😃 That's the approach I'm using but I'd rather not.

**Comment by Vadim-Zenin** [comment]
> 	thozza opened this issue on 4 Sep 2015
21 October 2018: no integration:

![20181021-175439-hubdocker-integration-limits](https://user-images.githubusercontent.com/3098552/47269808-b24d8d80-d55a-11e8-9aa5-5456b31def6d.png)

Please integrate GitLab, because a lot of Open Source communities and members will move from GitHub after Microsoft acquisition. It had happen with Novell and Suse. It will happen again.

[ EU commission approves acquisition of GitHub by Microsoft](http://europa.eu/rapid/press-release_IP-18-6155_en.htm)

**Comment by adawalli** [comment]
Github servers going down today are a good example of why we need more options (e.g., gitlab) for linked accounts!

**Comment by iMilnb** [comment]
Mandatory +1
That politics smelly smell...

**Comment by teodorescuserban** [comment]
TBH I asked for it as well but then , when you actually move to GitLab, you may realise it's far easier to just use their docker repository as well (and also their CI stuff) than hang on travis and docker hub.

And if you have an org account on github, travis, docker hub, well, you just get a subscription for gitlab and thats it.

I don't see why docker would add gitlab integration at this point. Probably because they would want to keep the current dockerhub customers?

**Comment by trickkiste** [comment]
When?

**Comment by dinigo** [comment]
Is this still on the roadmap? can someone link to the docker-hub issue-tracker in case they have one? Is there some positive confirmation that this is still being considered?

**Comment by rizerzero** [comment]
any update about this guys  ???

**Comment by Jonjoe** [comment]
bump

**Comment by zoobab** [comment]
It's 2019, this issue is 4 years old now. And Docker Hub is not Open Source Software, otherwise it would be fixed already :-)

**Comment by woodcockjosh** [comment]
@zoobab GitHub is not open source anymore than GitLab is

**Comment by 06kellyjac** [comment]
@woodcockjosh GitLab is *much* more open source than GitHub; it has an [open source core](https://gitlab.com/gitlab-org/gitlab-ce) and a very visible [enterprise code base](https://gitlab.com/gitlab-org/gitlab-ee) ...

Did you mean "GitHub is not open source anymore than Docker Hub"?

**Comment by dinigo** [comment]
If this is not in the roadmap anymore, please say so and close the issue so I can organize my workflow

**Comment by woodcockjosh** [comment]
> @woodcockjosh GitLab is _much_ more open source than GitHub; it has an [open source core](https://gitlab.com/gitlab-org/gitlab-ce) and a very visible [enterprise code base](https://gitlab.com/gitlab-org/gitlab-ee) ...
> 
> Did you mean "GitHub is not open source anymore than Docker Hub"?

No I meant that GitHub is not open source anymore than GitLab (because GitLab is more open sourced than GitHub)

**Comment by 06kellyjac** [comment]
Ahhhhhh I see what you mean.

**Comment by trickkiste** [comment]
I am a paying Docker Hub customer and I find it quite embarrassing, that you still don't support Gitlab. If you are not supporting them, because you consider their offerings a competition, you are achieving quite the opposite of what you might intend.

Because now I am forced to inspect their CI offerings in more detail in order to find out, if I can replace Docker Hub with Gitlab completely.

Suit yourselves! 

Regards,
Markus

**Comment by Yajo** [comment]
> Because now I am forced to inspect their CI offerings in more detail in order to find out, if I can replace Docker Hub with Gitlab completely.

I did it a long time ago, and TBH, Gitlab turned out to be way better than github and docker hub, its CI is awesome, and the image registry is fully integrated in the whole experience. I don't care anymore on this issue. Gitlab offers everything you'd need in one product. :blush:

**Comment by 06kellyjac** [comment]
I put my public containers on both just because Docker Hub is easier to browse for public images and is where the majority of people go but I specify how GitLab's registry is also faster to upload and download images (in my case) and waaaaaaaaaaay faster to build them.

If GitLab was to make it easier to search for public Docker images on their platform it would be in essence a drop in 100% improvement to Docker Hub minus the Docker branding that'd pull users that are unaware of GitLab's benefits.

**Comment by ovk** [comment]
Wow, still not GitLab integration, unbelievable.

**Comment by 0k00l** [comment]
bump

**Comment by PabloSzx** [comment]
up 😔

**Comment by dinigo** [comment]
I'm just leaving GitHub + dockerhub in favor of another better registry. Gitlab is the full stack I needed. Even though I'm a bit frustrated because of the lack of feedback in 3 years, I'm very grateful for the Docker team job

**Comment by ArjonBu** [comment]
Well, this sucks a lot. A better alternative would be quay.io. They support Gitlab too.

**Comment by rizerzero** [comment]
Still no updates on this?

**Comment by RAYs3T** [comment]
@ArjonBu Or just with GitLab itself. They provide a docker registry and build nodes.

**Comment by SuperElectron** [comment]
What's the status on gitlab support for this?

**Comment by robsonvn** [comment]
what joke

**Comment by robsonvn** [comment]
what a joke

**Comment by robsonvn** [comment]
moving away from docker hub

**Comment by robsonvn** [comment]
Obama was still the CC of USA when this ticket was opened.

**Comment by robsonvn** [comment]
4 days later the iPhone 6s and 6s Plus were released

**Comment by robsonvn** [comment]
Uptown Funk was the 1st song in the Billboard hot list

**Comment by robsonvn** [comment]
Facebook was still a thing

**Comment by robsonvn** [comment]
Everyone was discussing whether the dress was white and gold or blue and black

**Comment by robsonvn** [comment]
Left Shark was the meme of the year

**Comment by woodcockjosh** [comment]
omg spam

**Comment by marcoshuck** [comment]
it would be nice to have some sort of token registration instead of having to use an username and password for integrating dockerhub with gitlab.

**Comment by fasenderos** [comment]
Five years later, still not GitLab integration, embarrassing.

**Comment by ledleledle** [comment]
nowadays... 5 years passed. still no Gitlab integration

**Comment by jayjah** [comment]
greetings after 5 and a half, crazy that's still open ... But I think and hope(!) everyone found another, probably better solution :D

**Comment by yookoala** [comment]
The feature is still missing.

**Comment by 06kellyjac** [comment]
The feature is still missing.

**Comment by oldium** [comment]
Hey, bot, it is not our fault that nobody cares about real user problems.
